PyInstaller/building/imphook.py

```diff
@@ -12,18 +12,453 @@
+            `hook-{module_name}.py`, where `{module_name}` is the module hooked
+            by that script) to be cached.
+
+        See Also
+        ---------
+        `cache_hook_dir()`
+            Concrete details on hook script detection. and caching.
+        """
+
+        for hook_dir in hook_dirs:
+            self.cache_hook_dir(hook_dir)
+
+
+    def cache_hook_dir(self, hook_dir):
+        """
+        Cache all hook scripts in the passed directory.
+
+        **Order of caching is significant** with respect to hooks for the same
+        module, as the values of this dictionary are lists. Hooks for the same
+        module will be run in the order in which they are cached. Previously
+        cached hooks are always preserved rather than overidden.
+
+        By default, official hooks are cached _before_ user-defined hooks. For
+        modules with both official and user-defined hooks, this implies that the
+        former take priority over and hence will be loaded _before_ the latter.
+
+        Parameters
+        ----------
+        hook_dir : str
+            Absolute or relative path of the directory containing **hook
+            scripts** (i.e., Python scripts with filenames matching
+            `hook-{module_name}.py`, where `{module_name}` is the module hooked
+            by that script) to be cached. For convenience, this method applies
+            tilde and variable expansion to this path (e.g., a leading `~` will
+            be replaced by the absolute path of this user's home directory).
+        """
+
+        # Canonicalize this directory's path and validate its existence.
+        hook_dir = os.path.abspath(expand_path(hook_dir))
+        if not os.path.isdir(hook_dir):
+            raise FileNotFoundError(
+                'Hook directory "{}" not found.'.format(hook_dir))
+
+        # For each hook script in this directory...
+        hook_filenames = glob.glob(os.path.join(hook_dir, 'hook-*.py'))
+        for hook_filename in hook_filenames:
+            # Fully-qualified name of this hook's corresponding module,
+            # constructed by removing the "hook-" prefix and ".py" suffix.
+            module_name = os.path.basename(hook_filename)[5:-3]
+
+            # Lazily loadable hook object.
+            module_hook = ModuleHook(
+                module_graph=self.module_graph,
+                module_name=module_name,
+                hook_filename=hook_filename,
+                hook_module_name_prefix=self._hook_module_name_prefix,
+            )
+
+            # If at least one hook object has been cached for this module,
+            # append this hook object to the existing list of such objects.
+            if module_name in self:
+                self[module_name].append(module_hook)
+            # Else, create a new list containing this hook object.
+            else:
+                self[module_name] = [module_hook]
+
+
+    def uncache_modules(self, module_names):
```

Comment:
> I find the name missleadings. Why not just call it `remove`?

...like, because, and stuff? Very well. `remove_modules` it is.

> I'd prefer the function signature being `(self, *module_name)`

A sound suggestion. Variadic modules for the win. It shall be.